{"cells":[{"metadata":{},"cell_type":"markdown","source":"# **vLife Virtusa**\n\n# **Gene Similarity of Pandemics using BioPython**\n\nI am going to use Biopython to analyze the established sequence of the caronavirus. For documentation look here: BioPython\n\nSome facts regarding the virus itself: COVID-19\n\nIt is of the (+)ssRNA classification of viruses, which means it is a single stranded virus that can be directly translated into protein.\nThe actual virus is called SARS-CoV-2, Covid-19 is the name for the respiratory disease it causes.\n\n**Usecase Description**\n\nGene Similarity is to compare the SARS-CoV-2 virus sequence with the other pandemics and viruses like SARS, MERS, Civet-SARS,BAT-SARS, EBOLAV, camelus, plasmodium-malariae, hiv2, hedgehog and find the similarity scores between them. \n\n**SARS-Cov-2 sequence input - **\n\nThe SARS-Cov-2 sequence should be dropped in as the input and would be compared with other pandemics.\n\n**Description**\n\nThe first 100 sequence of COV-2 are pairwise compared with other pandemics using Bio python package pairwise and a score is determined based on the match.\nThen, the complete genome sequence of Cov-2 is compared with other pandemics and a score is determined based on the following:\n1. aligner.match_score = 1\n2. aligner.mismatch_score = -1\n3. aligner.open_gap_score = -0.5\n4. aligner.extend_gap_score = -0.1\n\nLater the score is divided by the total length of CoV-2 genome sequence (which is the input). based on which percentage of similarity is determined. \n\n\n**Input**\nThe Input used in the usecase is FASTA format DNA of COVID-19 isolate.\n\n**Input to the model**\n\nDNA FASTA file of SARS, MERS, Civet-SARS,BAT-SARS, EBOLAV, camelus, plasmodium-malariae, hiv2, hedgehog pandemics.\n\n**Dataset source**\nThe input dataset source - https://bigd.big.ac.cn/gwh/browse/virus/coronaviridae\n\nFasta file of SARS and MERS - https://www.kaggle.com/radenkovic/coronavirus-accession-sars-mers-cov2\n\nFasta file of Civet-SARS,BAT-SARS and EBOLAV - https://www.kaggle.com/akiator9/ebolav-vs-sarscov-vs-mers \n\nFasta file of camelus, plasmodium-malariae, hiv2 and hedgehog - https://www.kaggle.com/amansingh29/cameluscovfasta\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"import numpy as np\nimport pandas as pd\nimport re\nimport plotly.offline as py\nimport plotly.graph_objects as go\nfrom Bio import SeqIO\nimport os\nos.environ['QT_QPA_PLATFORM']='offscreen' \n\nimport matplotlib.pyplot as plt\n%matplotlib inline\nfrom colorama import Back, Style, Fore\n\ntry:\n    from Bio.Align.Applications import ClustalwCommandline\n    from Bio import Align, pairwise2\nexcept:\n    !pip install biopython\n    from Bio.Align.Applications import ClustalwCommandline\n    from Bio import Align, pairwise2","execution_count":5,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We use BioPython packages for the analysis"},{"metadata":{},"cell_type":"markdown","source":"# **DNA INPUT OF COV2 FASTA SEQUENCE**"},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"class Genesimilarity():\n    def genome(dna_seq):\n        with open(\"../input/coronavirus-accession-sars-mers-cov2/sars.fasta\", \"r\") as file:\n            sars_genome = file.read().split(\"genome\")[1].replace(\"\\n\", \"\")\n        with open(\"../input/coronavirus-accession-sars-mers-cov2/mers.fasta\", \"r\") as file:\n            mers_genome = file.read().split(\"genome\")[1].replace(\"\\n\", \"\")\n        sample_DNAseq = dna_seq\n        cov_genome = sample_DNAseq.seq\n        cov_genome = str(cov_genome)\n        with open(\"../input/ebolav-vs-sarscov-vs-mers/Civet-SARS.fasta\", \"r\") as file:\n            civet_sl_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/ebolav-vs-sarscov-vs-mers/BAT-SARS.fasta\", \"r\") as file:\n            bat_sl_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/ebolav-vs-sarscov-vs-mers/EBOLAV.fasta\", \"r\") as file:\n            ebola5_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/cameluscovfasta/camelus.fasta\", \"r\") as file:\n            camel_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/cameluscovfasta/plasmodium-malariae.fasta\", \"r\") as file:\n            malaria_genome = file.read().split(\"complete sequence\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/cameluscovfasta/hiv2.fasta\", \"r\") as file:\n            hiv2_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n        with open(\"../input/cameluscovfasta/hedgehog.fasta\", \"r\") as file:\n            hedgehog_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n\n        def gen_protein_seq(genome_str):\n            protein = {\"TTT\": \"F\", \"CTT\": \"L\", \"ATT\": \"I\", \"GTT\": \"V\",\n                   \"TTC\": \"F\", \"CTC\": \"L\", \"ATC\": \"I\", \"GTC\": \"V\",\n                   \"TTA\": \"L\", \"CTA\": \"L\", \"ATA\": \"I\", \"GTA\": \"V\",\n                   \"TTG\": \"L\", \"CTG\": \"L\", \"ATG\": \"M\", \"GTG\": \"V\",\n                   \"TCT\": \"S\", \"CCT\": \"P\", \"ACT\": \"T\", \"GCT\": \"A\",\n                   \"TCC\": \"S\", \"CCC\": \"P\", \"ACC\": \"T\", \"GCC\": \"A\",\n                   \"TCA\": \"S\", \"CCA\": \"P\", \"ACA\": \"T\", \"GCA\": \"A\",\n                   \"TCG\": \"S\", \"CCG\": \"P\", \"ACG\": \"T\", \"GCG\": \"A\",\n                   \"TAT\": \"Y\", \"CAT\": \"H\", \"AAT\": \"N\", \"GAT\": \"D\",\n                   \"TAC\": \"Y\", \"CAC\": \"H\", \"AAC\": \"N\", \"GAC\": \"D\",\n                   \"TAA\": \"STOP\", \"CAA\": \"Q\", \"AAA\": \"K\", \"GAA\": \"E\",\n                   \"TAG\": \"STOP\", \"CAG\": \"Q\", \"AAG\": \"K\", \"GAG\": \"E\",\n                   \"TGT\": \"C\", \"CGT\": \"R\", \"AGT\": \"S\", \"GGT\": \"G\",\n                   \"TGC\": \"C\", \"CGC\": \"R\", \"AGC\": \"S\", \"GGC\": \"G\",\n                   \"TGA\": \"STOP\", \"CGA\": \"R\", \"AGA\": \"R\", \"GGA\": \"G\",\n                   \"TGG\": \"W\", \"CGG\": \"R\", \"AGG\": \"R\", \"GGG\": \"G\"\n                   }\n            protein_seq = \"\"\n\n        # generating the protein seq\n            for i in range(0, len(genome_str) - (3 + len(genome_str) % 3), 3):\n                protein_seq += protein[genome_str[i:i + 3]]\n            return protein_seq\n\n        genomes = {\"SARS\": sars_genome, \"MERS\": mers_genome, \"COVID-19\": cov_genome, \"Civet_SL_CoV\": civet_sl_cov_genome,\n               \"Bat_SL_CoV\": bat_sl_cov_genome, \"Ebola\": ebola5_genome, \"Camel_CoV\": camel_cov_genome,\n               \"Malaria\": malaria_genome, \"HIV\": hiv2_genome, \"Hedgehog_CoV\": hedgehog_cov_genome}\n\n        def get_orfs(protein_seq):\n            orf_strands = []\n            for seq in protein_seq.split(\"STOP\"):\n                for nu in range(len(seq) - 1, -1, -1):\n                    if seq[nu] == \"M\":\n                        orf_strands.append(seq[nu:] + \"STOP\")\n        # search in original seq\n            patterns = \"|\".join(orf_strands)\n            res_seq = re.sub(patterns, lambda x: x.group(0).lower(), protein_seq)\n            for genome_name, genome in genomes.items():\n                get_orfs(gen_protein_seq(genome))\n\n        new1 = str(\"alilgning the first 100 nucleotides of COVID-19 genome with every other genomes:\\n\")\n        new2=[]\n        for genome_name, genome in genomes.items():\n            if genome_name != 'COVID-19':\n                new2.append((\"COVID-19 and {} genome\".format(genome_name)))\n                alignments = pairwise2.align.globalxx(cov_genome[:100], genome[:100])\n                new2.append((pairwise2.format_alignment(*alignments[0], full_sequences=True)))\n        new2 = '\\n'.join(new2)\n\n        aligner = Align.PairwiseAligner()\n        aligner.mode = 'global'\n        aligner.match_score = 1\n        aligner.mismatch_score = -1\n        aligner.open_gap_score = -0.5\n        aligner.extend_gap_score = -0.1\n\n        a = (\"Similarity scores between\\n\")\n        b=[]\n        for genome_name, genome in genomes.items():\n            if genome_name != 'COVID-19':\n                score = aligner.score(cov_genome, genome)\n                b.append(str(\"COVID-19 & {} genome sequences:\\t {} ({:.2f}%)\".format(genome_name, round(score),\n                                                                              100 * ((score) / len(cov_genome)))))\n        b = '\\n'.join(b)\n        return print(new1),print(new2),print(a),print(b)\n\n\ndna_seq = SeqIO.read(\"/kaggle/input/testdatagenome/NC_046965.genome.fasta\",\"fasta\")\nGenesimilarity.genome(dna_seq)","execution_count":7,"outputs":[{"output_type":"stream","text":"alilgning the first 100 nucleotides of COVID-19 genome with every other genomes:\n\nCOVID-19 and SARS genome\nACT-TTGAGCA-TTGATATATA--TAT---AT--ATATAT--C-ATACTCA-CCTT-G--C-CTTGTGCAAGAACCT-T-TTCTTAACA-AAACGGA-CTTATAG---TGCGTACG-GT---T-T-GCT-G--\n| | || ||   ||  | | ||  ||    |   | | |   | | || || || | |  | |||||   || |  | | ||| |  |  |||| || ||| ||    | | |  | ||   | | ||| |  \nA-TATT-AG--GTT--T-T-TACCTA-CCCA-GGA-A-A-AGCCA-AC-CAACC-TCGATCTCTTGT---AG-A--TCTGTTC-T--C-TAAAC-GAACTT-TA-AAAT-C-T--GTGTAGCTGTCGCTCGGC\n  Score=67\n\nCOVID-19 and MERS genome\n-ACTTTGAGCAT-TGATATA--TATA---TATA-T-A-TATCATACTCAC-C-TTGCCT-TGTGCAAGAACCTTT--TCTTAACAA-AACGGA-CTTATAG-T----GCGTAC--G--GTTT-GC-T---G-\n | ||| |  |  ||| |||  | |    |||  | | | ||   | | | | ||  || | ||| |||| ||||  | ||      ||| || |||| |  |    ||   |  |  |||| || |   | \nGA-TTT-A--A-GTGA-ATAGCT-T-GGCTAT-CTCACT-TC---C-C-CTCGTT--CTCT-TGC-AGAA-CTTTGAT-TT-----TAAC-GAACTTA-A-ATAAAAGC---CCTGTTGTTTAGCGTATCGT\n  Score=68\n\nCOVID-19 and Civet_SL_CoV genome\nACTTTGAGCATTGATATATATATATATATATCATACTCA-CCTT-G--C-CTTGTGCAAGAACCT-T-TTCT-TAA-C-AA----AACGGA--CT-TA-TAGT-G-CG-TACGG-T---T---T-G--C---T--G-----\n|     |||                      || || || || | |  | |||||   || |  | | |||| ||| | ||    ||   |  || |  |||| | || | ||| |   |   | |  |   |  |     \nA-----AGC----------------------CA-AC-CAACC-TCGATCTCTTGT---AG-A--TCTGTTCTCTAAACGAACTTTAA---AATCTGT-GTAGTTGTCGCT-CGGCTGCATGCCTAGTGCACCTACGCAGTA\n  Score=59\n\nCOVID-19 and Bat_SL_CoV genome\nACT-TTGAGCA-TTGATATATA--TAT---AT--ATATAT--C-ATACTCA-CCTT-G--C-CTTGTGCAAGAACCT-T-TTCTTAACA-AAACGGA-CTTATAG---TGCGTACG-GT---T-T-GCT-G--\n| | || ||   ||  | | ||  ||    |   | | |   | | || || || | |  | |||||   || |  | | |||||     |||| || ||| ||    | | |  | ||   | | ||| |  \nA-TATT-AG--GTT--T-T-TACCTA-CCCA-GGA-A-A-AGCCA-AC-CAACC-TCGATCTCTTGT---AG-A--TCTGTTCTT----TAAAC-GAACTT-TA-AAAT-C-T--GTGTGGCTGTCGCTTGGC\n  Score=67\n\nCOVID-19 and Ebola genome\nACTTTGAG-CATTG-AT-ATATATATAT-ATATATC-ATACTC-ACC-TTGCCTTGTGCAAGA-AC-CTTTTC-T-TAA-C-AA-AAC---G-G-AC---TTA-TAG-TG---CG-TACGGT--TTGC---T-------G----\n |   | | ||    |  | | | | |  | | |   | |    |   ||   || |   ||  |  |||||  | |   | || |||   | | |    ||| ||  |    |  | |  |  |||    |       |    \n-C---G-GACA---CA-CA-A-A-A-A-GA-A-A--GA-A---GA--ATT---TT-T---AG-GA-TCTTTT-GTGT--GCGAATAACTATGAGGA-AGATTAATA-AT-TTTC-CT-C--TCATTG-AAATTTATATCGGAAT\n  Score=56\n\nCOVID-19 and Camel_CoV genome\n-ACTTTGAGCAT-TGATATA--TATA---TATA-T-A-TATCATACTCAC-C-TTGCCT-TGTGCAAGAACCTTT--TCTTAACAA-AACGGA-CTTA--TA---G---TGC--GT--ACG-GT-TTGCTG\n | ||| |  |  ||| |||  | |    |||  | | | ||   | | | | ||  || | ||| |||| ||||  | ||      ||| || ||||  ||   |   ||   ||  | | || ||| | \nGA-TTT-A--A-GTGA-ATAGCT-T-GGCTAT-CTCACT-TC---C-C-CTCGTT--CTCT-TGC-AGAA-CTTTGAT-TT-----TAAC-GAACTTAAATAAAAGCCCTG-TTGTTTA-GCGTATTG-T-\n  Score=69\n\nCOVID-19 and Malaria genome\nACTT-TGA-G-CATTG-ATATATATATATATA--TATCATACTCAC-CT--TGCCT-TG-TGC-AAGA--A--CCTTTT-CT-T--AA-CAA-A-AC----GG-ACT-TATAGTGCG-T--ACGG-T-TTGCTG-\n |   | | | |  |  | | || | | | |   ||| | | | |  |   || |  || ||  ||||  |  ||   | |  |  || ||| | ||    || | | | | || || |  | || | || |   \n-C--CT-ACGGC--T-CA-A-AT-T-T-T-T-GCTAT-A-A-T-A-GC-AATG-C-GTGATG-GAAGAGTATACC---TAC-ATCCAAGCAACACACCCTAGGGA-TGT-T-GT-CGGTGTA-GGGTATT-C--A\n  Score=65\n\nCOVID-19 and HIV genome\nACTTTG-A-GC--ATTGATATATATAT-A-TAT--AT--AT--CATACTC-ACCT-TGCCTTGTGCAAGAACCTT-T-T-CT-TA-ACAAAACG-G-ACT-----TA-TA---G-T-GCGTACG-G--TT-TG-CT---G-\n    || | |   | || | | | ||  | ||   |   |   ||||    |    | ||   |   || | |   | | |  || | |||| | | |       || ||   | | | |  |  |  || |  ||   | \n----TGCAAG-GGA-TG-T-T-T-TA-CAGTA-GGA-GGA-GACATA---GA---AT-CC---T---AG-A-C--ATATAC-CTAGA-AAAA-GAGGA--AGGGATAATACCAGATTG-G--C-AGAATTAT-ACTCATGG\n  Score=59\n\nCOVID-19 and Hedgehog_CoV genome\n-ACTTTGAGCATTGAT-AT-ATATAT--ATA--TATA---TCATACTCA-CC-TTGC-CTTGTGCAAGAAC-CTT-TTCTTAACAAAACGGA-CTTATAGT-----GCG--T-A--CGGTTTG-CT-G----\n |  || |  |   |  |  | |||   |||  |||    || | |||  || || | |||||   ||||| ||| || ||     ||| || |||| | |     ||   | |  |  |||| || |    \nGA--TT-A--A---A-GA-GA-ATA-GCATAGCTAT-CCCTC-T-CTC-TCCGTT-CTCTTGT---AGAACTCTTGTT-TT-----AAC-GAACTTA-A-TTAAAAGC-CCTTATAC--TTTGTCTAGGTAG\n  Score=68\n\nSimilarity scores between\n\nCOVID-19 & SARS genome sequences:\t 14103 (49.42%)\nCOVID-19 & MERS genome sequences:\t 14190 (49.72%)\nCOVID-19 & Civet_SL_CoV genome sequences:\t 13979 (48.98%)\nCOVID-19 & Bat_SL_CoV genome sequences:\t 14031 (49.17%)\nCOVID-19 & Ebola genome sequences:\t 9827 (34.43%)\nCOVID-19 & Camel_CoV genome sequences:\t 14198 (49.75%)\nCOVID-19 & Malaria genome sequences:\t 135 (0.47%)\nCOVID-19 & HIV genome sequences:\t 5853 (20.51%)\nCOVID-19 & Hedgehog_CoV genome sequences:\t 14456 (50.65%)\n","name":"stdout"},{"output_type":"execute_result","execution_count":7,"data":{"text/plain":"(None, None, None, None)"},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"# **Conclusion -** \n\nThe DNA sequence of a Cov-2 entered is pairwise compared with other pandemics like SARS, MERS, Civet-SARS,BAT-SARS, EBOLAV, camelus, plasmodium-malariae, hiv2, hedgehog pandemics and similarity score between COVID-19 Genome and other Genome is determined."}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}