{"cells":[{"metadata":{},"cell_type":"markdown","source":"# **vLife Virtusa**\n\n# **Genome Analysis and finding frequency of Protien using BioPython**\n\nI am going to use Biopython to analyze the established sequence of the caronavirus. For documentation look here: BioPython\n\nSome facts regarding the virus itself: COVID-19\n\nIt is of the (+)ssRNA classification of viruses, which means it is a single stranded virus that can be directly translated into protein.\nThe actual virus is called SARS-CoV-2, Covid-19 is the name for the respiratory disease it causes.\n\n**Usecase Description**\n\nGenome Analysis is performed with BioPython, The FASTA DNA sequence or Protein sequence is considered as an input.\n\n**DNA sequence input - **\n\nDNA sequence is transcribed (T (for Thymine) is replaced with U (for Uracil)) to mRNA which is then translated to Proteins, then functional protiens are extracted from the protein sequences(which is protien of length>20).\nWe analyse the largest protein sequence and we obtain the following:\n1. The Aminoacid frequency is then obtained from the largest protein sequence found.\n2. IsoElectric point - The isoelectric point, is the pH at which a molecule carries no net electrical charge or is electrically neutral in the statistical mean\n3. Aromaticity - In organic chemistry, aromaticity is a property of cyclic, planar structures with a ring of resonance bonds that gives increased stability compared to other geometric or connective arrangements with the same set of atoms.\n4. Molecular weight - The molecular mass is the mass of a given molecule: it is measured in daltons. Different molecules of the same compound may have different molecular masses because they contain different isotopes of an element.\n5. Secondary structure composition.\n\n**Protein sequence input**\n\nProtein sequence is considered as an input and then functional protiens are extracted from the protein sequences(which is protien of length>20).\nWe analyse the largest protein sequence and we obtain the following:\n1. The Aminoacid frequency is then obtained from the largest protein sequence found.\n2. IsoElectric point - The isoelectric point, is the pH at which a molecule carries no net electrical charge or is electrically neutral in the statistical mean\n3. Aromaticity - In organic chemistry, aromaticity is a property of cyclic, planar structures with a ring of resonance bonds that gives increased stability compared to other geometric or connective arrangements with the same set of atoms.\n4. Molecular weight - The molecular mass is the mass of a given molecule: it is measured in daltons. Different molecules of the same compound may have different molecular masses because they contain different isotopes of an element.\n5. Secondary structure composition.\n\n**Dataset**\nThe Dataset used in the usecase is FASTA format DNA or Protein of COVID-19 isolate.\n\n**Dataset source**\nThe dataset source - https://bigd.big.ac.cn/gwh/browse/virus/coronaviridae"},{"metadata":{"trusted":true},"cell_type":"code","source":"import pandas as pd\nimport os\nfrom flask import Flask, render_template, request\nimport matplotlib.pyplot as plt\nfrom Bio import SeqIO\nfrom Bio.SeqRecord import SeqRecord\nfrom werkzeug.utils import secure_filename\nimport re\n###New section ############\n\nimport Bio\nimport operator\nfrom Bio.Seq import Seq\nfrom Bio.Alphabet import generic_dna, generic_protein, generic_rna\nfrom Bio.Blast import NCBIWWW\nfrom Bio.SeqUtils.ProtParam import ProteinAnalysis\nfrom Bio import SeqIO\nfrom Bio import pairwise2\n\nimport pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\n\n\namino_acids_mapping = {v: k for k, v in Bio.SeqUtils.IUPACData.protein_letters_3to1.items()}\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We use BioPython packages for the analysis"},{"metadata":{},"cell_type":"markdown","source":"# **DNA INPUT**"},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"class genome():\n    def genome_frequency(dnaseq):\n        sample_DNAseq = dnaseq\n        sample_DNA = sample_DNAseq.seq\n        sample_DNA_a = sample_DNA[0:100]\n        bb = ('A', 'T', 'C', 'G', 'Y', 'W', 'U')\n        for x in str(sample_DNA_a):\n            if x not in bb:\n                result_value = False\n                print(\"You have entered a Protein file\")\n                break\n            else:\n                result_value = True\n                print(\"You have entered a DNA file\")\n                break\n        if result_value == True:\n            mRNA_sample = sample_DNA.transcribe()\n\n            Amino_Acid = mRNA_sample.translate()\n            lenofamino = len(Amino_Acid)\n            lenmRna = len(mRNA_sample)\n            from Bio.Data import CodonTable\n            condon = CodonTable.unambiguous_rna_by_name['Standard']\n\n            Proteins_sample = Amino_Acid.split('*')\n\n            Protein_dataframe = pd.DataFrame(Proteins_sample)\n            Protein_dataframe.describe()\n\n            def conv(item):\n                return len(item)\n\n            def to_str(item):\n                return str(item)\n\n            Protein_dataframe['sequence_string'] = Protein_dataframe[0].apply(to_str)\n            Protein_dataframe['length'] = Protein_dataframe[0].apply(conv)\n            Protein_dataframe.rename(columns={0: \"sequence of protein\"}, inplace=True)\n\n            functional_proteins = Protein_dataframe.loc[Protein_dataframe['length'] >= 20]\n            lenfunprotein = len(functional_proteins)\n\n            functional_proteins.describe()\n            indexprotein = Protein_dataframe['length'].idxmax()\n\n            for i in Proteins_sample[:]:\n                if len(i) < 20:\n                    Proteins_sample.remove(i)\n            Proteins_sample\n            largest_prot = max(Proteins_sample, key=len)\n            lenlarget_prot = len(largest_prot)\n            indexprot = Proteins_sample.index(largest_prot)\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n\n            MolecularWeight = []\n            amino_freq = []\n            aromaticity = []\n            IsoElectric = []\n            for i in Proteins_sample[:]:\n                test = ProteinAnalysis(str(i))\n                MolecularWeight.append(test.molecular_weight())\n                amino_freq.append(test.count_amino_acids())\n                aromaticity.append(test.aromaticity())\n                IsoElectric.append(test.isoelectric_point())\n\n            aromaticity_d = aromaticity[indexprot]\n            aromaticity_d = str(round(aromaticity_d, 2))\n            IsoElectric_d = IsoElectric[indexprot]\n            IsoElectric_d = str(round(IsoElectric_d, 2))\n\n            MolecularWeight = pd.DataFrame(data=MolecularWeight, columns=[\"Molecular Weights\"])\n            MolecularWeight['Molecular Weights'].max()\n\n            MolecularWeight.idxmax()\n            from collections import Counter\n            res = Counter(largest_prot)\n\n            alpha_result = []\n            listToStr = ' '.join([str(elem) for elem in largest_prot])\n            bb = ('A', 'E', 'G', 'K', 'L', 'M')\n            for x in listToStr:\n                if x in bb:\n                    # print(x)\n                    alpha_result.append(x)\n\n            alpha_result = ' '.join([str(elem) for elem in alpha_result])\n\n            largest_prot_str = ''.join([str(elem) for elem in largest_prot])\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n            X = ProteinAnalysis(largest_prot_str)\n            sec_struc = X.secondary_structure_fraction()\n            mol_weight = X.molecular_weight()\n            mol_weight = str(round(mol_weight, 2))\n            mRNA_sample = str(mRNA_sample)\n            Amino_Acid = str(Amino_Acid)\n            largest_prot = str(largest_prot)\n            return {\"Length of RNA \":lenmRna},{\"Length of Amino Acids \":lenofamino},{\"Length of functional protiens \":lenfunprotein},{\"Length of the largest protein \":lenlarget_prot},{\"Aromaticity \":aromaticity_d}, {\"Isoelectricity \":IsoElectric_d}, {\"Frequency of Amino Acids \":res}, {\"Secondary Structure composition \":sec_struc}, {\"Molecular weight \": mol_weight}\n        else:\n            sample_prot = dnaseq\n            Proteins_sample = sample_prot.seq\n            largest_prot = Proteins_sample\n            lenlarget_prot = len(largest_prot)\n            indexprot = Proteins_sample.index(largest_prot)\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n\n            MolecularWeight = []\n            amino_freq = []\n            aromaticity = []\n            IsoElectric = []\n            for i in Proteins_sample[:]:\n                test = ProteinAnalysis(str(i))\n                MolecularWeight.append(test.molecular_weight())\n                amino_freq.append(test.count_amino_acids())\n                aromaticity.append(test.aromaticity())\n                IsoElectric.append(test.isoelectric_point())\n\n            aromaticity_d = aromaticity[indexprot]\n            aromaticity_d = str(round(aromaticity_d, 2))\n            IsoElectric_d = IsoElectric[indexprot]\n            IsoElectric_d = str(round(IsoElectric_d, 2))\n\n            MolecularWeight = pd.DataFrame(data=MolecularWeight, columns=[\"Molecular Weights\"])\n            MolecularWeight['Molecular Weights'].max()\n\n\n            MolecularWeight.idxmax()\n\n            from collections import Counter\n            res = Counter(largest_prot)\n\n            alpha_result = []\n            listToStr = ' '.join([str(elem) for elem in largest_prot])\n            bb = ('A', 'E', 'G', 'K', 'L', 'M')\n            for x in listToStr:\n                if x in bb:\n                    # print(x)\n                    alpha_result.append(x)\n\n            alpha_result = ' '.join([str(elem) for elem in alpha_result])\n\n            largest_prot_str = ''.join([str(elem) for elem in largest_prot])\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n            X = ProteinAnalysis(largest_prot_str)\n            sec_struc = X.secondary_structure_fraction()\n\n            mol_weight = X.molecular_weight()\n            mol_weight = str(round(mol_weight, 2))\n            largest_prot = str(largest_prot)\n            return {\"Length of the largest protein \":lenlarget_prot}, {\"Aromaticity \":aromaticity_d}, {\"Isoelectricity \":IsoElectric_d}, {\"Frequency of Amino Acids \":res}, {\"Secondary Structure composition \":sec_struc}, {\"Molecular weight \": mol_weight}\n\n\n\ndna_seq = SeqIO.read(\"/kaggle/input/testdatagenome/NC_046965.genome.fasta\",\"fasta\")\ngenome.genome_frequency(dna_seq)","execution_count":6,"outputs":[{"output_type":"stream","text":"You have entered a DNA file\n","name":"stdout"},{"output_type":"execute_result","execution_count":6,"data":{"text/plain":"({'Length of RNA ': 28539},\n {'Length of Amino Acids ': 9513},\n {'Length of functional protiens ': 46},\n {'Length of the largest protein ': 2697},\n {'Aromaticity ': '0.12'},\n {'Isoelectricity ': '6.58'},\n {'Frequency of Amino Acids ': Counter({'L': 224,\n           'F': 159,\n           'K': 155,\n           'R': 106,\n           'V': 230,\n           'G': 154,\n           'S': 181,\n           'E': 104,\n           'A': 160,\n           'I': 131,\n           'P': 112,\n           'T': 165,\n           'D': 170,\n           'Q': 93,\n           'C': 97,\n           'N': 164,\n           'M': 59,\n           'H': 61,\n           'Y': 144,\n           'W': 28})},\n {'Secondary Structure composition ': (0.3396366332962551,\n   0.2265480163144234,\n   0.20281794586577678)},\n {'Molecular weight ': '304733.82'})"},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"# **Protein Input**"},{"metadata":{"trusted":true},"cell_type":"code","source":"class genome():\n    def genome_frequency(dnaseq):\n        sample_DNAseq = dnaseq\n        sample_DNA = sample_DNAseq.seq\n        sample_DNA_a = sample_DNA[0:100]\n        bb = ('A', 'T', 'C', 'G', 'Y', 'W', 'U')\n        for x in str(sample_DNA_a):\n            if x not in bb:\n                result_value = False\n                print(\"You have entered a Protein file\")\n                break\n            else:\n                result_value = True\n                print(\"You have entered a DNA file\")\n                break\n        if result_value == True:\n            mRNA_sample = sample_DNA.transcribe()\n\n            Amino_Acid = mRNA_sample.translate()\n            lenofamino = len(Amino_Acid)\n            lenmRna = len(mRNA_sample)\n            from Bio.Data import CodonTable\n            condon = CodonTable.unambiguous_rna_by_name['Standard']\n\n            Proteins_sample = Amino_Acid.split('*')\n\n            Protein_dataframe = pd.DataFrame(Proteins_sample)\n            Protein_dataframe.describe()\n\n            def conv(item):\n                return len(item)\n\n            def to_str(item):\n                return str(item)\n\n            Protein_dataframe['sequence_string'] = Protein_dataframe[0].apply(to_str)\n            Protein_dataframe['length'] = Protein_dataframe[0].apply(conv)\n            Protein_dataframe.rename(columns={0: \"sequence of protein\"}, inplace=True)\n\n            functional_proteins = Protein_dataframe.loc[Protein_dataframe['length'] >= 20]\n            lenfunprotein = len(functional_proteins)\n\n            functional_proteins.describe()\n            indexprotein = Protein_dataframe['length'].idxmax()\n\n            for i in Proteins_sample[:]:\n                if len(i) < 20:\n                    Proteins_sample.remove(i)\n            Proteins_sample\n            largest_prot = max(Proteins_sample, key=len)\n            lenlarget_prot = len(largest_prot)\n            indexprot = Proteins_sample.index(largest_prot)\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n\n            MolecularWeight = []\n            amino_freq = []\n            aromaticity = []\n            IsoElectric = []\n            for i in Proteins_sample[:]:\n                test = ProteinAnalysis(str(i))\n                MolecularWeight.append(test.molecular_weight())\n                amino_freq.append(test.count_amino_acids())\n                aromaticity.append(test.aromaticity())\n                IsoElectric.append(test.isoelectric_point())\n\n            aromaticity_d = aromaticity[indexprot]\n            aromaticity_d = str(round(aromaticity_d, 2))\n            IsoElectric_d = IsoElectric[indexprot]\n            IsoElectric_d = str(round(IsoElectric_d, 2))\n\n            MolecularWeight = pd.DataFrame(data=MolecularWeight, columns=[\"Molecular Weights\"])\n            MolecularWeight['Molecular Weights'].max()\n\n            MolecularWeight.idxmax()\n            from collections import Counter\n            res = Counter(largest_prot)\n\n            alpha_result = []\n            listToStr = ' '.join([str(elem) for elem in largest_prot])\n            bb = ('A', 'E', 'G', 'K', 'L', 'M')\n            for x in listToStr:\n                if x in bb:\n                    # print(x)\n                    alpha_result.append(x)\n\n            alpha_result = ' '.join([str(elem) for elem in alpha_result])\n\n            largest_prot_str = ''.join([str(elem) for elem in largest_prot])\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n            X = ProteinAnalysis(largest_prot_str)\n            sec_struc = X.secondary_structure_fraction()\n            mol_weight = X.molecular_weight()\n            mol_weight = str(round(mol_weight, 2))\n            mRNA_sample = str(mRNA_sample)\n            Amino_Acid = str(Amino_Acid)\n            largest_prot = str(largest_prot)\n            return {\"Length of RNA \":lenmRna},{\"Length of Amino Acids \":lenofamino},{\"Length of functional protiens \":lenfunprotein},{\"Length of the largest protein \":lenlarget_prot},{\"Aromaticity \":aromaticity_d}, {\"Isoelectricity \":IsoElectric_d}, {\"Frequency of Amino Acids \":res}, {\"Secondary Structure composition \":sec_struc}, {\"Molecular weight \": mol_weight}\n        else:\n            sample_prot = dnaseq\n            Proteins_sample = sample_prot.seq\n            largest_prot = Proteins_sample\n            lenlarget_prot = len(largest_prot)\n            indexprot = Proteins_sample.index(largest_prot)\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n\n            MolecularWeight = []\n            amino_freq = []\n            aromaticity = []\n            IsoElectric = []\n            for i in Proteins_sample[:]:\n                test = ProteinAnalysis(str(i))\n                MolecularWeight.append(test.molecular_weight())\n                amino_freq.append(test.count_amino_acids())\n                aromaticity.append(test.aromaticity())\n                IsoElectric.append(test.isoelectric_point())\n\n            aromaticity_d = aromaticity[indexprot]\n            aromaticity_d = str(round(aromaticity_d, 2))\n            IsoElectric_d = IsoElectric[indexprot]\n            IsoElectric_d = str(round(IsoElectric_d, 2))\n\n            MolecularWeight = pd.DataFrame(data=MolecularWeight, columns=[\"Molecular Weights\"])\n            MolecularWeight['Molecular Weights'].max()\n\n\n            MolecularWeight.idxmax()\n\n            from collections import Counter\n            res = Counter(largest_prot)\n\n            alpha_result = []\n            listToStr = ' '.join([str(elem) for elem in largest_prot])\n            bb = ('A', 'E', 'G', 'K', 'L', 'M')\n            for x in listToStr:\n                if x in bb:\n                    # print(x)\n                    alpha_result.append(x)\n\n            alpha_result = ' '.join([str(elem) for elem in alpha_result])\n\n            largest_prot_str = ''.join([str(elem) for elem in largest_prot])\n\n            from Bio.SeqUtils.ProtParam import ProteinAnalysis\n            X = ProteinAnalysis(largest_prot_str)\n            sec_struc = X.secondary_structure_fraction()\n\n            mol_weight = X.molecular_weight()\n            mol_weight = str(round(mol_weight, 2))\n            largest_prot = str(largest_prot)\n            return {\"Length of the largest protein \":lenlarget_prot}, {\"Aromaticity \":aromaticity_d}, {\"Isoelectricity \":IsoElectric_d}, {\"Frequency of Amino Acids \":res}, {\"Secondary Structure composition \":sec_struc}, {\"Molecular weight \": mol_weight}\n\n\n\ndna_seq = SeqIO.read(\"/kaggle/input/prot-file/1P9U_B.Protein.faa\",\"fasta\")\ngenome.genome_frequency(dna_seq)\n","execution_count":7,"outputs":[{"output_type":"stream","text":"You have entered a Protein file\n","name":"stdout"},{"output_type":"execute_result","execution_count":7,"data":{"text/plain":"({'Length of the largest protein ': 302},\n {'Aromaticity ': '0.0'},\n {'Isoelectricity ': '5.24'},\n {'Frequency of Amino Acids ': Counter({'S': 31,\n           'G': 31,\n           'L': 26,\n           'R': 11,\n           'K': 12,\n           'M': 10,\n           'A': 13,\n           'Q': 8,\n           'P': 8,\n           'V': 30,\n           'E': 19,\n           'C': 6,\n           'I': 13,\n           'Y': 14,\n           'N': 25,\n           'W': 3,\n           'D': 7,\n           'H': 6,\n           'T': 16,\n           'F': 13})},\n {'Secondary Structure composition ': (0.32781456953642385,\n   0.31456953642384106,\n   0.22516556291390727)},\n {'Molecular weight ': '33094.13'})"},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"# **Conclusion -** \n\nThe sequence entered is analysed using BioPython packages and the frequency of Amino acids present in the largest protein sequence along with the Aromaticity, Isoelectric point, Molecular weight and secondary structure composition is rendered as the output"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}